### PR TITLE
Added support for background webserver for containerized applications

### PR DESCRIPTION
**Pull request by tameem92**

### CODE REVIEW

**Comment by tameem92** [comment]
@Strech Opened new PR following our conversation in https://github.com/Strech/sidekiq-prometheus-exporter/pull/3.

**Comment by Strech** [comment]
@tameem92 Hey, thanks for the PR, I will throw some questions soon 😉

**Comment by Strech** [comment]
@tameem92 Can I ask you a bit of more examples to understand the problem you try to solve?

Because if we will do that it will narrow down the solution to a specific problem. Instead, I think you should do an `ENTRYPOINT`

`Dockerfile`
```dockerfile
COPY entrypoint.sh /entrypoint.sh
ENTRYPOINT /entrypoint.sh
```

`entrypoint.sh`
```sh
#!/bin/sh

bundle exec rackup -d -p9292 -o0.0.0.0 metrics.ru
```

`metrics.ru`
```ruby
require 'sidekiq'
require 'sidekiq/prometheus/exporter'

Sidekiq.configure_client do |config|
  config.redis = {url: 'redis://<your-redis-host>:6379/0'}
end

run Sidekiq::Prometheus::Exporter.to_app
```

This will allow you to have anything you want in the background without a need to create a specific solution in the gem.

WDYT?

**Comment by tameem92** [comment]
@Strech Let's say I have a Ruby on Rails app that I am deploying in a Docker container to AWS Elastic Beanstalk. In the Dockerfile we use one of the web servers like Puma or Unicorn. We ideally want one docker container with Rail's application running. Now in order to deploy Sidekiq in Elastic Beanstalk, I need to deploy a new container with a similar code base which only runs Sidekiq which has no web server. 

Now the Sidekiq container entry point is:

```
# run.sh 
bundle install
bundle exec sidekiq
```

In this case, our entry point is the Sidekiq worker but I also want a web server running in order to expose the Sidekiq metrics. This applies to environments like Elastic Beanstalk or Kuberentes. If you look at the source code of Gitlab, they had to implement something similar and add a web server just for exposing sidekiq metrics. 

One way to accomplish this is to have supervisord running in the container and run two processes: 

1) bundle exec rackup -d -p9292 -o0.0.0.0 metrics.ru
2) bundle exec sidekiq

But I preferred the approach of having a simple function in the gem that runs the background server. With this approach, all I have to do is add that to my sidekiq config. And just run the command "bundle exec sidekiq" 

Let me know what you think? Thanks.

**Comment by Strech** [comment]
@tameem92 in my current job we are also using Docker containers. And if I get what you want – you want to have a Sidekiq metrics within your Sidekiq container, but you `ENTRYPOINT` is already a Sidekiq process?

And you don't want to spawn a process from the same `ENTRYPOINT` file (a new web server with `-d` option), but instead, you want to have a co-running code in a thread within your Sidekiq? Or Rails?

Because if you want, you also can register an exporter and then if you have `sidekiq/web` enabled it will be there.

1. I understand that you don't want to run more than 1 process in a container, but what is preventing you from having code you add in you initializers/lib/etc of your Rails application if you want to run exporter there?
2. Can you give me an example you mention from the GitLab?
3. Can you clarify from 2 containers you have 1st – Rails, 2nd – Sidekiq, in which you want to add Exporter?

**Comment by tameem92** [comment]
@Strech That's correct. I want the metrics to be available in the Sidekiq Docker Container and instead of spawning a second process in my ENTRYPOINT file. Spawning a second process in the file using "-d" is totally a valid solution, but I thought it would be good if the gem just had a function to do that. This gets rid of having an ENTRYPOINT file and you can just have a CMD at the end too. 

I definitely noticed the solution of having it available on ``` Sidekiq/web ``` but with AWS infrastructure available with in a VPC, I don't want to have the dashboard ```sidekiq/web``` available publicly and its firewalled to whitelist a few IP addresses. 

This is the issue from Gitlab: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/13082
They added options for a webrick server for exporting sidekiq metrics so its part of Gitlab and people won't have to manually add it in the ```ENTRYPOINT``` file.

To sum it up, having the function I added or adding a web server running in daemon mode in the ``` ENTRYPOINT ``` file are both functionally equivalent. A few advantages with having it available in the Gem and in the README. 

1) A person using the gem won't have to do extra tweaks in their Docker configs for Sidekiq. 
2) From the README it's clear if they have a container that doesn't run a web server they can use this solution. 

If you think adding this to the ENTRYPOINT file is a better solution. I would just go ahead and update the README with that information regarding this case.

**Comment by Strech** [comment]
@tameem92 Thanks for your feedback! I would like to ask you a bit more. So in your `sidekiq.rb` why don't you want to do something like this?

```ruby
require 'sidekiq'
require 'sidekiq/prometheus/exporter'

Sidekiq.configure_client do |config|
  config.redis = {url: 'redis://<your-redis-host>:6379/0'}
end

Thread.new do
  Rack::Handler::WEBrick.run(Sidekiq::Prometheus::Exporter.to_app, Host: '0.0.0.0', Port: 3001, AccessLog: [])
end
```
It 2 lines longer than `run Sidekiq::Prometheus::Exporter.start_metrics_server(host, port)` but it's more flexible and it allows you do do any middleware you want.

Also it's quite confusing why I should use `run` from `Rack` even if it's `Webrick.run` in a `Thread`. And you include some middlewared I think overwhelming, like logging and so on.

Hence I think we should include it into README but not into a gem itself. It's too specific and opens many question, like: why this set of middleware, why can't I use block as [Webrick allows me to do](https://github.com/rack/rack/blob/4fe7d538c2ded6c5f4e55a12ad9bc10c994b49cb/lib/rack/handler/webrick.rb#L27-L37) and so on.

But I'm happy to think about it one more time if we will see someone else need ❤️

**Comment by tameem92** [comment]
@Strech That makes sense. Will just add it to README and close off the PR. Thanks.

**Comment by Strech** [comment]
@tameem92 I will appreciate if you can enhance the readme it might help others 👏
